normalizer_lib.py

```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/env python
+"""Library of language-agnostic FST rewrite rules to normalize text."""
+
+import unicodedata
+from pynini import *
+from config import *
+
+LANGUAGE = af
+
+GRAPHEMES = LANGUAGE.GRAPHEMES
+
+INITIAL_PUNCTUATION = LANGUAGE.INITIAL_PUNCTUATION
+FINAL_PUNCTUATION = LANGUAGE.FINAL_PUNCTUATION
+OTHER_PUNCTUATION = union(r"\[", r"\]")
+PUNCTUATION = union(INITIAL_PUNCTUATION, FINAL_PUNCTUATION, OTHER_PUNCTUATION)
+NUMBERS = union("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
+SPACE = acceptor(" ")
+
+SIGMA_STAR = union(*("[{}]".format(i) for i in range(1, 256))
+                   ).optimize().closure()
+
+
+# Step 1: Remove all extra whitespace between words
+
+REMOVE_EXTRA_WHITESPACE = transducer(SPACE, "")
+
+DO_REMOVE_EXTRA_WHITESPACE = cdrewrite(
+    REMOVE_EXTRA_WHITESPACE,
+    "",
+    SPACE,
+    SIGMA_STAR)
+
+
+# Step 2: Language-specific formatting fixes
+
+DO_LANGUAGE_SPECIFIC_PREPROCESSING = LANGUAGE.LANGUAGE_SPECIFIC_PREPROCESSING
+
+
+# Step 3: Apply NFC unicode normalize
+# really slow when done this way,
+# super fast when I just call unicodedata.normalize
+# on the string at the beginning
+# may remove later
+#def unicode_normalize(string: str) -> Fst:
+#    "Normalizes unicode to NFC normalization."
+#    nfc_normalized = unicodedata.normalize("NFC", string)
+#    normalizer = transducer(string, nfc_normalized).optimize()
+#    return cdrewrite(normalizer, "", "", SIGMA_STAR).optimize()
+
+
+# Step 4: Discard examples not associated with a pronunciation
+# e.g. "how are you today товарищ?" -> "how are you today?"
+
+SENTENCE = (INITIAL_PUNCTUATION.ques + (GRAPHEMES.plus | NUMBERS.plus) + FINAL_PUNCTUATION.ques +
+            (SPACE.plus + GRAPHEMES.plus).star + FINAL_PUNCTUATION.ques)
+
+
+def pass_only_valid(string: str) -> str:
+    "Allows only valid tokens of the language to pass through."
```

Comment:
Also, could you beef up the docstring comment to explain that we're checking individual tokens and not full sentences?

Additionally, for the docstrings, can you add "returns:" and "args:"?  Here's an example: https://google.github.io/styleguide/pyguide.html#244-decision

Reply:
Sounds good. I tweaked the one-sentence description and added `args` and `returns` thus:

```
"""Replaces invalid tokens in a sentence with <UNK>, keeps the rest.

Args:
    string: A line from a corpus.

Returns:
    The line, but where invalid tokens have been replaced by <UNK>.
"""
```